### DISCUSSION TITLE
[讨论]缓存的新思路

### DISCUSSION THREAD

**Discussion by mkanako** [IrineSistiana/mosdns#62] [General]
首次请求->检查缓存当然没有->请求上游->得到结果返回并写入缓存（假设ttl=5分钟）

5分钟后再次请求->检查缓存有一个过期的记录但是别删->直接返回（注意，返回的时候把ttl改得很小，比如就30秒吧）->同时去请求上游得到新的结果更新进缓存

这样除了首次之外，后续都直接从缓存返回非常快，即便偶尔拿到过期的记录但由于ttl被改得很小，可以让客户端尽快再次请求拿到最新结果，最大限度的发挥缓存的作用

目前来讲memory cache，默认size是1024条，我很怀疑真的有人能让缓存同时超过1024条吗，不到这个数目LRU根本没活干，它就是个摆设，如果按照这思路实现，定时清除器可以去掉了，LRU很快就有活干了

但是redis cache好像就难搞了🤣

@IrineSistiana 怎么看😎

**Comment by CalmLong** [comment]
对于 CDN 的资源会不会出现第一次点击打不开，第二次点击才打开的情况，毕竟 CDN 的 TTL 本身就很短

**Reply by mkanako** [reply]
有可能，这是最大的问题😬

**Comment by IrineSistiana** [comment]
反对。lazy update 的方式用在 dns 上有些不妥。

不过主动更新是可以的，类似 unbound 的 prefetch https://nlnetlabs.nl/documentation/unbound/unbound.conf/

比如主动更新缓存里最常用的前 10% 的域名。

**Reply by mkanako** [reply]
哈哈，果然不太行😝

**Comment by jamiesun** [comment]
即使是30秒的TTL也会对客户端造成很大困扰， 这就像程序出现一个本来应该抛出去的异常， 你给他忽略了， 客户端说不准啥时候暴雷， 这类似于违反快速失败原则

**Comment by mkanako** [comment]
纯属雷同

已有按照这个思路实现的dns服务器
https://github.com/LEXUGE/dcompass/

AdGuardHomes上也有这样的[Feature Request]
https://github.com/AdguardTeam/AdGuardHome/issues/2145

@jamiesun 
@IrineSistiana

**Reply by mkanako** [reply]
我不太懂go，想自己实现这个插件出来，本以为拿原本的cache插件稍微改改就行了

结果，[拿到过期缓存直接返回并请求上游去更新缓存]这一步完全搞不定🤣

**Reply by IrineSistiana** [reply]
参考“最终一致性系统”。DNS 就是一个 “最终一致性系统”。所以我一直坚持 ttl 只能变小不能变大，也就是说只能 prefetch，不能 lazy update。

~~虽然`ttl`插件已经违背了，但就这样吧~~

> 并请求上游去更新缓存]这一步

你可以让用户指定一个更新用的 插件 tag。然后去执行它。

**Reply by mkanako** [reply]
嗯，插件的我再研究研究

这里呢我主要讲究实用性，这玩意主要拿来自用的，实用第一~~原则什么的随意~~😝

**Reply by LEXUGE** [reply]
我是dcompass的作者，关于lazy update的问题，其实就是为了缓解极差网络环境下且遇到大量inbounding请求的情况。不过失效后第一个请求的程序必然会得到outdated的结果，但是一般程序都是resilient的，加上outdated的responses也不是不可用。
prefetch的话如果存在大量突发请求，那后续prefetch更新的时候也会给自己造成压力。
当然lazy也不文明（
考虑后续refactor

**Comment by zhaibin18** [comment]
cache应该是共享的，比如A请求了www.google.com，返回结果放在cache缓存。

×如果在ttl的时间内，B请求www.google.com，就应该立即返回缓存的结果，不管是不是带有cookie的等选项；
×如果超过了ttl的时间，B请求www.google.com，立即返回cache的内容，并且再去上游重新查询更新结果。

上面的机制应该是unbond的模式，我使用了这种模式大概有1年的时间，正常的家庭上网、看剧、玩游戏，没有什么异常,DNS的响应速度也很快。dcompass的问题是返回的ttl永远都不变，而不是随上游的DNS服务器进行更新的。
